token-metadata/program/tests/verify.rs

```diff
@@ -0,0 +1,1929 @@
+#![cfg(feature = "test-bpf")]
+
+pub mod utils;
+
+use mpl_token_metadata::{
+    error::MetadataError,
+    instruction::{
+        builders::VerifyBuilder, DelegateArgs, InstructionBuilder, MetadataDelegateRole, VerifyArgs,
+    },
+    pda::{find_metadata_delegate_record_account, find_token_record_account},
+    state::{Collection, CollectionDetails, Creator, TokenStandard},
+};
+use num_traits::FromPrimitive;
+use solana_program::native_token::LAMPORTS_PER_SOL;
+use solana_program_test::*;
+use solana_sdk::{
+    instruction::InstructionError, signature::Keypair, signer::Signer, transaction::Transaction,
+    transaction::TransactionError,
+};
+use utils::*;
+
+// Note: at the time these tests were created, the only Metadata delegates that have been
+// implemented are `Collection`, `Update`, and `ProgrammableConfig`.  We have tested each of these
+// cases.
+
+// Also at this time, a collection parent NFT cannot have a token standard of
+// `ProgrammableNonFungible`.  This means that using the new delegate handler, the only Token
```

Comment:
Where is this restriction enforced?

Reply:
https://github.com/metaplex-foundation/metaplex-program-library/blob/master/token-metadata/program/src/assertions/collection.rs#L139

I think its just outdated and will fix as part of this PR.

Reply:
Fixed `assert_master_edition` but still going to add some tests.

Reply:
Added tests for NFT/pNFT collections

And a test that makes sure in the case of pNFT, that Utility Delegate cannot verify or unverify.